Adds proper URL decoding to Riak's REST API

jtuple opened the pull request:
Addresses bz://617 and bz://1003.

New sane behavior where Riak URL decodes buckets/keys sent as URLs and Headers over REST API. Behavior is configurable through 'http_url_decoding' app.config parameter where 'on' defaults to new behavior, and missing/anything else defaults to compatibility mode where links are decoded but buckets/keys are not.

This pull request also includes migration support that can help users migrate an existing cluster with encoded bucket/keys to one without. This is handled in riak_kv_encoding_migrate.erl.

jtuple committed:
Add test for pending URL encoding changes to REST API.

-- The compat test checks current behavior and should pass.
-- The sane test checks for the new behavior and should fail.

jtuple committed:
Add support for proper REST URL encoding behavior.

-- Add riak_kv app value: http_url_encoding. When set to 'on',
   riak correctly decodes URI encoded bucket/keys. Otherwise,
   riak uses previous behavior: decode links, but not bucket/keys.

-- All tests in rest_url_encoding_test now pass.

jtuple committed:
Add test for 'X-Riak-URL-Encoding' header.

jtuple committed:
Add support for 'X-Riak-URL-Encoding' header in compat mode.

jtuple committed:
Provide support to migrate URL encoded buckets/keys.

jtuple committed:
Fix link walking issue with new URL encoding behavior.

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,228 @@
```
This file needs a license header.

beerriot commented on test/rest_url_encoding_test.erl:
```diff
@@ -0,0 +1,241 @@
```
This file should probably have a license header as well.  Some in test/ do, others don't.

beerriot commented on test/rest_url_encoding_test.erl:
```diff
@@ -0,0 +1,241 @@
+    %% Retrieve link header and check that it is singly encoded, corresponding
+    %% to a decoded internal link.
+    {ok, {_, Headers, _}} = 
+        httpc:request(get, {"http://localhost:8091/riak/rest_links/compat",
```
change `"http://localhost:8091` to `URL ++ "`, and in other places below (or test fails for non-dev1:8091 nodes)

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,228 @@
+            io:format("Cluster is empty. No migration needed.~n", []),
+            {ok, empty};
+        _ ->
+            MR = [RC:mapred(Bucket, [map_check(), reduce_check()])
```
This MapReduce query is going to read every object stored in the cluster.  Since check_object/1 only needs the bucket and key to make its decision, this request could be converted to reduce-only (no map phase), and not have to read any objects from disk.  The reduce phase will receive a list of bucket/key pairs, and can operate directly on them.  Most of an example of doing this can be found here: https://github.com/basho/riak_kv/commit/b5e969008343e466f245a2faadf1f364b7cba915

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,228 @@
+
+%% Perform first phase of migration: copying encoded values to
+%% unencoded equivalents.
+migrate_objects() ->
```
IMO, it would be useful to have this log the list of objects it converts, whether it do so in a file, or by returning the list of converted keys, or by just printing them on the console.  It would help with troubleshooting later know have a record of which things converted, and how.

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,228 @@
+    end.
+
+%% Perform second phase of migration: delete objects with encoded buckets/keys.
+delete_migrated_objects() ->
```
Same as migrate_objects: logging would be a good idea here.

jtuple committed:
Add license header and remove hardcoded test URLs.

jtuple committed:
Update URL encoding migration code.

-- Favor reduce over map to minimize I/O.
-- Check for doubly encoded buckets/keys.

jtuple commented:
Forgot to mention this earlier: my last set of commits should address all review items.

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,282 @@
+    end.
+
+%% Returns a list of URL encoded objects needing migration.
+get_encoded_objects(RC) ->
```
Let's export a zero-arity form of this function as well, for people to use  if they hit the `{custom, []}` cases above.  I might suggest naming it `get_encoded_keys` as well, just to not confuse people when they don't get `riak_object` records back.

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,282 @@
+    case MR of
+        {ok, [unsafe]} ->
+            unsafe;
+        _ ->
```
This will return `safe` if MR is an error tuple.

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,282 @@
+map_safe({error, notfound}, _, _) ->
+    [];
+map_safe(_, _, _) ->
+    [unsafe].
```
It might be more user-friendly to return `[{riak_object:bucket(O), riak_object:key(O)}]` here.  Skip the reduce phase in `check_safe/2`, and match on `[]` for the safe case, but return the list of unsafe keys for a user to comb through, without having to repeat this dig manually.

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,282 @@
+
+%% Returns true if any value is double (or more) encoded.
+check_double_encoding(EObjs) ->
+    lists:any(fun(Name) ->
```
It might be more user-friendly to `lists:filter` here instead, so `check_cluster/2` can return a list of the keys it doesn't know how to migrate, instead of just saying that some exist.

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,282 @@
+    [];
+map_rewrite_encoded(RO, _, _) ->
+    {_, _, B2, K2} = decode_object(RO),
+    copy_object(RO, B2, K2),
```
Probably worth checking the return of `copy_object/3`.  If it's not `ok`, maybe return the bucket/key as the result, so the user can check up on failures?

beerriot commented on src/riak_kv_encoding_migrate.erl:
```diff
@@ -0,0 +1,282 @@
+%% Perform second phase of migration: delete objects with encoded buckets/keys.
+delete_migrated_objects(EObjs) ->
+    {ok, RC} = riak:local_client(),
+    [RC:delete(B, K) || {B, K} <- EObjs],
```
Probably worth checking that all deletes returned `ok`.

jtuple committed:
Make URL migration code more helpful under failure.

jtuple commented:
Latest commit should address the remaining issues. There are a few cases were we can't really recover gracefully, so I'm just letting a match fail (eg. map_unsafe when the MR job fails), but otherwise I believe we handle all cases where it makes sense to provide the user with "bad objects" that can be dealt with manually.

beerriot commented:
+1 merge
